File: platform/Swing/sweeteditor/src/main/java/com/qiplat/sweeteditor/copilot/InlineSuggestionController.java
package com.qiplat.sweeteditor.copilot;

import com.qiplat.sweeteditor.EditorTheme;
import com.qiplat.sweeteditor.SweetEditor;
import com.qiplat.sweeteditor.core.adornment.PhantomText;
import com.qiplat.sweeteditor.event.*;

import java.awt.event.KeyEvent;
import java.util.List;

/**
 * Controller managing inline suggestion lifecycle: PhantomText injection,
 * event listening, Tab/Esc interception, and action bar display.
 */
public class InlineSuggestionController {

    private final SweetEditor editor;
    private InlineSuggestionActionBar actionBar;
    private InlineSuggestion currentSuggestion;
    private InlineSuggestionListener listener;
    private boolean showing = false;

    private final EditorEventListener<TextChangedEvent> textListener = e -> autoDismiss();
    private final EditorEventListener<CursorChangedEvent> cursorListener = e -> autoDismiss();

    public InlineSuggestionController(SweetEditor editor) {
        this.editor = editor;
    }

    public void show(InlineSuggestion suggestion) {
        if (suggestion == null || suggestion.text == null || suggestion.text.isEmpty()) return;

        // Dismiss previous suggestion if any
        if (showing) {
            clearPhantomText();
        }

        currentSuggestion = suggestion;
        showing = true;

        // Inject phantom text
        editor.setLinePhantomTexts(suggestion.line,
                List.of(new PhantomText(suggestion.column, suggestion.text)));
        editor.flush();

        // Subscribe to events for auto-dismiss
        editor.subscribe(TextChangedEvent.class, textListener);
        editor.subscribe(CursorChangedEvent.class, cursorListener);

        // Show action bar
        if (actionBar == null) {
            actionBar = new InlineSuggestionActionBar(editor, editor.getTheme(),
                    this::accept, this::dismiss);
        }
        actionBar.showAtCursor();
    }

    public void accept() {
        if (!showing) return;
        InlineSuggestion suggestion = currentSuggestion;

        clearAndUnsubscribe();

        // Insert text at suggestion position
        if (suggestion != null) {
            editor.getEditorCore().setCursorPosition(suggestion.line, suggestion.column);
            editor.insertText(suggestion.text);
        }

        if (listener != null && suggestion != null) {
            listener.onSuggestionAccepted(suggestion);
        }
    }

    public void dismiss() {
        if (!showing) return;
        InlineSuggestion suggestion = currentSuggestion;

        clearAndUnsubscribe();

        if (listener != null && suggestion != null) {
            listener.onSuggestionDismissed(suggestion);
        }
    }

    public boolean isShowing() {
        return showing;
    }

    public void setListener(InlineSuggestionListener listener) {
        this.listener = listener;
    }

    public void applyTheme(EditorTheme theme) {
        if (actionBar != null) {
            actionBar.applyTheme(theme);
        }
    }

    /**
     * Intercepts Tab (accept) and Escape (dismiss) key events.
     * @return true if the key was consumed
     */
    public boolean handleKeyEvent(int keyCode) {
        if (!showing) return false;
        if (keyCode == KeyEvent.VK_TAB) {
            accept();
            return true;
        }
        if (keyCode == KeyEvent.VK_ESCAPE) {
            dismiss();
            return true;
        }
        return false;
    }

    public void updatePosition(float cursorX, float cursorY, float cursorHeight) {
        if (actionBar != null && showing) {
            actionBar.updatePosition(cursorX, cursorY, cursorHeight);
        }
    }

    private void autoDismiss() {
        if (!showing) return;
        InlineSuggestion suggestion = currentSuggestion;
        clearAndUnsubscribe();
        if (listener != null && suggestion != null) {
            listener.onSuggestionDismissed(suggestion);
        }
    }

    private void clearAndUnsubscribe() {
        showing = false;
        clearPhantomText();
        currentSuggestion = null;

        editor.unsubscribe(TextChangedEvent.class, textListener);
        editor.unsubscribe(CursorChangedEvent.class, cursorListener);

        if (actionBar != null) {
            actionBar.dismiss();
        }
    }

    private void clearPhantomText() {
        if (currentSuggestion != null) {
            editor.setLinePhantomTexts(currentSuggestion.line, List.of());
            editor.flush();
        }
    }
}

File: platform/Flutter/sweeteditor/lib/copilot/inline_suggestion_controller.dart
part of '../sweeteditor.dart';

/// Callback for inline suggestion action bar Accept/Dismiss interaction.
abstract class InlineSuggestionActionCallback {
  void onAcceptClicked();
  void onDismissClicked();
}

class InlineSuggestionOverlayState {
  const InlineSuggestionOverlayState({
    required this.x,
    required this.y,
    required this.cursorHeight,
  });

  final double x;
  final double y;
  final double cursorHeight;
}

/// Manages inline suggestion lifecycle: phantom text injection, event subscriptions,
/// Tab/Esc key interception, and action bar state.
class InlineSuggestionController implements InlineSuggestionActionCallback {
  InlineSuggestionController({required EditorSession session})
    : _session = session,
      _eventBus = session.eventBus;

  final EditorSession _session;
  final EditorEventBus _eventBus;

  InlineSuggestion? _currentSuggestion;
  InlineSuggestionListener? _listener;
  bool _showing = false;
  bool _suppressAutoDismiss = false;
  double _cachedCursorX = 0;
  double _cachedCursorY = 0;
  double _cachedCursorHeight = 0;
  EditorOverlayBinding<InlineSuggestionOverlayState>? _overlayBinding;
  StreamSubscription<TextChangedEvent>? _textChangedSub;
  StreamSubscription<CursorChangedEvent>? _cursorChangedSub;
  StreamSubscription<ScrollChangedEvent>? _scrollChangedSub;

  void _onTextChanged(TextChangedEvent _) => _autoDismiss();

  void _onCursorChanged(CursorChangedEvent _) => _autoDismiss();

  void _onScrollChanged(ScrollChangedEvent _) {
    if (_showing) {
      _overlayBinding?.update(_buildOverlayState());
    }
  }

  void setListener(InlineSuggestionListener? listener) {
    _listener = listener;
  }

  void bindOverlay(EditorOverlayBinding<InlineSuggestionOverlayState>? binding) {
    _overlayBinding = binding;
  }

  bool get isShowing => _showing;

  void show(InlineSuggestion suggestion) {
    if (_showing) _clearQuietly();
    _currentSuggestion = suggestion;
    _injectPhantomText(suggestion);

    final cursor = _session.renderModel.cursor;
    _cachedCursorX = cursor.position.x;
    _cachedCursorY = cursor.position.y;
    _cachedCursorHeight = cursor.height;

    _showing = true;
    _overlayBinding?.show(_buildOverlayState());
    _subscribeEvents();
    _session.requestFlush();
  }

  void accept() {
    if (_currentSuggestion == null) return;
    final suggestion = _currentSuggestion!;
    _withSuppressedAutoDismiss(() {
      _unsubscribeEvents();
      _session.editorCore?.clearPhantomTexts();
      final pos = core.TextPosition(suggestion.line, suggestion.column);
      _session.editorCore?.replaceText(
        pos.line,
        pos.column,
        pos.line,
        pos.column,
        suggestion.text,
      );
      _showing = false;
      _overlayBinding?.hide();
      _currentSuggestion = null;
      _session.requestFlush();
    });
    _listener?.onSuggestionAccepted(suggestion);
  }

  void dismiss() {
    if (_currentSuggestion == null) return;
    final suggestion = _currentSuggestion!;
    _withSuppressedAutoDismiss(() {
      _unsubscribeEvents();
      _session.editorCore?.clearPhantomTexts();
      _session.requestFlush();
      _showing = false;
      _overlayBinding?.hide();
      _currentSuggestion = null;
    });
    _listener?.onSuggestionDismissed(suggestion);
  }

  /// Handle key codes. Returns true if consumed.
  /// Tab -> accept, Escape -> dismiss.
  bool handleKeyCode(int keyCode) {
    if (!_showing) return false;
    if (keyCode == 9) {
      accept();
      return true;
    }
    if (keyCode == 27) {
      dismiss();
      return true;
    }
    return false;
  }

  void updatePosition(double cursorX, double cursorY, double cursorHeight) {
    _cachedCursorX = cursorX;
    _cachedCursorY = cursorY;
    _cachedCursorHeight = cursorHeight;
    if (_showing) {
      _overlayBinding?.update(_buildOverlayState());
    }
  }

  @override
  void onAcceptClicked() => accept();

  @override
  void onDismissClicked() => dismiss();

  void dispose() {
    _unsubscribeEvents();
    _overlayBinding = null;
    _currentSuggestion = null;
    _showing = false;
  }

  void _autoDismiss() {
    if (_suppressAutoDismiss) return;
    dismiss();
  }

  void _clearQuietly() {
    _withSuppressedAutoDismiss(() {
      _unsubscribeEvents();
      _session.editorCore?.clearPhantomTexts();
      _showing = false;
      _overlayBinding?.hide();
      _currentSuggestion = null;
    });
  }

  void _withSuppressedAutoDismiss(void Function() action) {
    _suppressAutoDismiss = true;
    try {
      action();
    } finally {
      _suppressAutoDismiss = false;
    }
  }

  void _injectPhantomText(InlineSuggestion suggestion) {
    _session.editorCore?.clearPhantomTexts();
    _session.editorCore?.setBatchLinePhantomTexts({
      suggestion.line: [
        core.PhantomText(column: suggestion.column, text: suggestion.text),
      ],
    });
  }

  InlineSuggestionOverlayState _buildOverlayState() {
    return InlineSuggestionOverlayState(
      x: _cachedCursorX,
      y: _cachedCursorY,
      cursorHeight: _cachedCursorHeight,
    );
  }

  void _subscribeEvents() {
    _textChangedSub ??= _eventBus.on<TextChangedEvent>().listen(_onTextChanged);
    _cursorChangedSub ??=
        _eventBus.on<CursorChangedEvent>().listen(_onCursorChanged);
    _scrollChangedSub ??=
        _eventBus.on<ScrollChangedEvent>().listen(_onScrollChanged);
  }

  void _unsubscribeEvents() {
    _textChangedSub?.cancel();
    _textChangedSub = null;
    _cursorChangedSub?.cancel();
    _cursorChangedSub = null;
    _scrollChangedSub?.cancel();
    _scrollChangedSub = null;
  }
}
